flytekit/types/structured/structured_dataset.py

```diff
@@ -737,10 +737,20 @@ async def async_to_literal(
             #       return StructuredDataset(uri=uri)
             if python_val.dataframe is None:
                 uri = python_val.uri
+                file_format = python_val.file_format
+
+                # Check the user-specified uri
                 if not uri:
                     raise ValueError(f"If dataframe is not specified, then the uri should be specified. {python_val}")
                 if not ctx.file_access.is_remote(uri):
                     uri = await ctx.file_access.async_put_raw_data(uri)
+
+                # Check the user-specified file_format
+                # When users specify file_format for a StructuredDataset, the file_format information must be retained.
+                # For details, please refer to https://github.com/flyteorg/flyte/issues/6096.
+                if file_format != GENERIC_FORMAT:
+                    sdt.format = file_format
```

Comment:
Should the file format always be copied over?

```suggestion
                sdt.format = file_format
```

Reply:
Hi @thomasjpfan,

Thanks for your suggestion. After tracing the code logic, I think it's totally fine to copy it. I'll remove the redundant condition.

Reply:
Hi @thomasjpfan,

I think we can't always copy it over. I've come up with the following use case:
```python
@task
def modify_format(sd: Annotated[StructuredDataset, {}, "task-format"]) -> StructuredDataset: 
    return sd

sd = StructuredDataset(uri="s3://my-s3-bucket/df.parquet", file_format="user-format")
sd2 = modify_format(sd=sd)
```

In this case, we expect `sd2.file_format` to be `task-format` (as shown in `Annotated`), not `user-format`. If we always use the user-specified `file_format`, the information set in `Annotated` will be missing.

Considering `sdt.format` can be set [here](https://github.com/flyteorg/flytekit/blob/3260ddfea32b0a772cf002da643a02299be1021a/flytekit/types/structured/structured_dataset.py#L703-L709), would it be good to do provide a stricter condition as follows?
```python
if file_format != GENERIC_FORMAT and sdt.format == GENERIC_FORMAT:
    sdt.format = file_format
```

Reply:
Yup that is good with me.

Reply:
Thanks a lot for reviewing this.